### ISSUE TITLE
Better pattern matching code for generic deriving large sum types

### ISSUE THREAD

**Issue by ajnsit** [purescript/purescript#3948]
I just ran into the issue described in [generic deriving's JS output for long sum types](https://github.com/purescript/purescript/issues/3475). The JS output was extremely surprising. I understand that there might be bigger issues involved in changing the data representation, however I believe that there are low hanging fruit here that will ameliorate the problem somewhat.

Instead of generating code like this (from @hdgarrood's [gist explaining generics](https://try.purescript.org/?gist=c99f9ebf3bec311d1a5065f3fa4bfaa1&session=3ee7a943-c4b9-ef38-ec0f-048a13031350) ) -

```purescript
repToPokémonStatus = case _ of
  Inl (Inl counter) -> Asleep counter
  Inl (Inr (severity :*: counter)) -> Poisoned severity counter
  Inr _ -> Paralyzed
```

We can generate -

```purescript
repToPokémonStatus = case _ of
  Inl x -> case x of
    Inl counter -> Asleep counter
    Inr (severity :*: counter) -> Poisoned severity counter
  Inr _ -> Paralyzed
```

This should reduce the size of the code generated by quite a bit, and make it a *lot* more performant.

**Comment by ajnsit** [comment]
I can start taking a look at this, unless someone thinks that such a PR would not be appropriate, or has a better suggestion.

**Comment by garyb** [comment]
Maybe a more useful approach in general would be to recognise the former case and codegen it as if it were the latter case?

**Comment by hdgarrood** [comment]
That’s still O(n) in the worst case. I’d rather balance the Generic structure so that it’s O(log n), and look into generating optimized patterns after that.

**Comment by f-f** [comment]
@hdgarrood wouldn't that make writing code with Generics more involved, as you'd have to take care of more cases? (i.e. right now you can lean on the fact that the tree only grows in one direction. So this would also make it a fairly breaking change)

**Comment by ajnsit** [comment]
@garyb that would be very useful, however I'm it might be much harder to do that for general data structures. In this limited case however, we know that there are only 2 cases and we control the entire codepath.

@hdgarrood It's also help reign in the code size, which really balloons up currently. I ran into this in a production app where a simple module with a data structure with ~80 variants generated 1MB of JS code. Removing the generic instance reduced the size to 25kb. In the absence of a perfect fix that might take much longer, can we have a quick fix that solves 80% of the problem?

**Comment by hdgarrood** [comment]
@f-f yes, it would be a breaking change. However, I think the fact that converting to/from Generic is O(n) where n is the number of constructors is a serious enough problem that it deserves consideration anyway. I think it’s possible that a good chunk of existing code won’t need modification.

@ajnsit I generally prefer to avoid “quick fixes” since they usually incur more effort than just doing it properly from the start.

Personally I would prefer to investigate optimizing pattern matching generally (what @garyb said), or to investigate switching Generic to use a balanced representation, before looking into this. However, both of those are likely going to be larger tasks. If the approach described here turns out to be very straightforward then I’m happy to consider it.

**Comment by ajnsit** [comment]
@hdgarrood Sure I will take a look at this to see how much of a change it is.

One thing though, AFAIK currently the pattern matching is more like O(2^n). For example, for 3 constructors, the worst case pattern match involves matching against LLL, LLR, LRL, LRR, RLL, RLR, RRL, RRR. This change will make it O(n).

**Comment by hdgarrood** [comment]
Oh right yes; I think the amount of code generated is more or less quadratic at the moment, since we do take the balancing into account, but the amount of code generated for the nth case is more or less proportional to n.

**Comment by garyb** [comment]
I think there probably is a reasonably quick fix here that isn't going to be anywhere near as involved as the other suggestions, and it shouldn't be too complicated or "harmful" to include, so yeah I'd be happy with that for the time being too. A lot of the code around pattern matching and its decomposition into JS/CoreImp is pretty thorny.

**Comment by radrow** [comment]
Just wanted to point out that the presented idea doesn't trivially solve the issue if the match looks like
```
let repToPokémonStatus = case _ of
      Inl (Inl counter) -> Asleep counter
      Inr _ -> Paralyzed
      Inl (Inr (severity :*: counter)) -> Poisoned severity counter
      a -> Burned

in repToPokémonStatus (Inl Someotherthing) == Burned && repToPokémonStatus NeitherInrNotInl == Burned
```
(ofc, we can always duplicate the code with `Burned`)

also the same problem with multi-arg functions, where
```
let f True False = 1
    f False _ = 2
    f _ True = 3
    
    g = f True
in g True 
```
should return `3`. 

Btw I really appreciate that this issue got highlighted. Hoping this to get fixed soon.
